scripts/update_function_docs/update_function_docs.go

```diff
@@ -0,0 +1,294 @@
+			semver.Compare(patchVersion, latestPatchVersion) == 1 {
+			latestPatchVersion = patchVersion
+			lang = segments[len(segments)-3]
+		}
+	}
+	if latestPatchVersion == "" || lang == "" {
+		return fmt.Errorf("could not find matching tag for release branch")
+	}
+	fr.Language = lang
+	fr.LatestPatchVersion = latestPatchVersion
+	return nil
+}
+
+// readDocPaths and set FunctionPath and ExamplePaths
+func (fr *functionRelease) readDocPaths() error {
+	executablePath, err := os.Executable()
+	if err != nil {
+		return err
+	}
+	repoBase := filepath.Dir(filepath.Dir(filepath.Dir(executablePath)))
+	fr.FunctionPath = filepath.Join(repoBase, "functions", fr.Language, fr.FunctionName)
+	examplesPath := filepath.Join(repoBase, "examples")
+
+	if !dirExists(fr.FunctionPath) {
+		fr.FunctionPath = filepath.Join(repoBase, "contrib", "functions", fr.Language, fr.FunctionName)
+		examplesPath = filepath.Join(repoBase, "contrib", "examples")
+		if !dirExists(fr.FunctionPath) {
+			return fmt.Errorf("function doc paths not found %s", fr.FunctionPath)
+		}
+	}
+	if err = fr.parseMetadata(examplesPath); err != nil {
+		return err
+	}
+	return nil
+}
+
+// parseMetadata from metadata.yaml and set ExamplePaths
+func (fr *functionRelease) parseMetadata(examplesPath string) error {
+	type metadata struct {
+		ExamplePackageUrls []string `yaml:"examplePackageURLs"`
+	}
+	if fr.FunctionPath == "" {
+		return fmt.Errorf("expected FunctionPath in parseMetadata")
+	}
+
+	metadataPath := filepath.Join(fr.FunctionPath, "metadata.yaml")
+	var md metadata
+	yamlFile, err := ioutil.ReadFile(metadataPath)
+	if err != nil {
+		return err
+	}
+
+	err = yaml.Unmarshal(yamlFile, &md)
+	if err != nil {
+		return err
+	}
+	for _, exampleURL := range md.ExamplePackageUrls {
+		segments := strings.Split(exampleURL, "/")
+		exampleName := segments[len(segments)-1]
+		examplePath := filepath.Join(examplesPath, exampleName)
+		if !dirExists(examplePath) {
+			return fmt.Errorf("example dir does not exist: %s", examplePath)
+		}
+		fr.ExamplePaths = append(fr.ExamplePaths, examplePath)
+	}
+	return nil
+}
+
+// updateDoc with the latest version in place at specified path
+func (fr *functionRelease) updateDoc(path string) error {
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	versionGroup := `(unstable|v\d*\.\d*\.\d*|v\d*\.\d*)`
+	// replace tags with patch e.g. apply-setters:v1.0.1, apply-setters/v1.0.1
+	tagPattern := regexp.MustCompile(
+		fmt.Sprintf(`%s(:|/)%s`, fr.FunctionName, versionGroup))
+	contents = tagPattern.ReplaceAll(contents,
+		[]byte(fmt.Sprintf(`%s${1}%s`, fr.FunctionName, fr.LatestPatchVersion)))
+	// replace url with minor e.g. https://catalog.kpt.dev/apply-setters/v1.0
+	urlPattern := regexp.MustCompile(
+		fmt.Sprintf(`https://catalog\.kpt\.dev/%s/%s`, fr.FunctionName, versionGroup))
+	contents = urlPattern.ReplaceAll(contents,
+		[]byte(fmt.Sprintf(`https://catalog.kpt.dev/%s/%s`, fr.FunctionName, fr.MinorVersion)))
+
```

Comment:
What about updating
- from `kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval-simple` to `kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval-simple@kubeval/v0.2.1`
- from `kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval-simple/v0.2.0` to `kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval-simple@kubeval/v0.2.1`

Reply:
The case I have intentionally accounted for is
- `kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval-simple@kubeval/v0.2` to `kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/kubeval-simple@kubeval/v0.2.1`

Which is handled by the `(:|/)` capturing group.

I was not aware that there are other cases that need to be handled. Do you have examples for the other cases we need to handle?

Reply:
I did a bit of refactoring and added a `replaceKptPackages` method to handle this case